src/libstd/collections/hash/raw/generic.rs

```diff
@@ -0,0 +1,127 @@
+use super::bitmask::BitMask;
+use super::EMPTY;
+use core::{mem, ptr};
+
+// Use the native word size as the group size. Using a 64-bit group size on
+// a 32-bit architecture will just end up being more expensive because
+// shifts and multiplies will need to be emulated.
+#[cfg(any(
+    target_pointer_width = "64",
+    target_arch = "aarch64",
+    target_arch = "x86_64",
+))]
+type GroupWord = u64;
+#[cfg(all(
+    target_pointer_width = "32",
+    not(target_arch = "aarch64"),
+    not(target_arch = "x86_64"),
+))]
+type GroupWord = u32;
+
+pub type BitMaskWord = GroupWord;
+pub const BITMASK_SHIFT: u32 = 3;
+pub const BITMASK_MASK: GroupWord = 0x8080808080808080u64 as GroupWord;
+
+/// Helper function to replicate a byte across a `GroupWord`.
+#[inline]
+fn repeat(byte: u8) -> GroupWord {
+    let repeat = byte as GroupWord;
+    let repeat = repeat | repeat.wrapping_shl(8);
+    let repeat = repeat | repeat.wrapping_shl(16);
+    // This last line is a no-op with a 32-bit GroupWord
+    repeat | repeat.wrapping_shl(32)
+}
+
+/// Abstraction over a group of control bytes which can be scanned in
+/// parallel.
+///
+/// This implementation uses a word-sized integer.
+pub struct Group(GroupWord);
+
+// We perform all operations in the native endianess, and convert to
+// little-endian just before creating a BitMask. The can potentially
+// enable the compiler to eliminate unnecessary byte swaps if we are
+// only checking whether a BitMask is empty.
+impl Group {
+    /// Number of bytes in the group.
+    pub const WIDTH: usize = mem::size_of::<Self>();
+
+    /// Returns a full group of empty bytes, suitable for use as the initial
+    /// value for an empty hash table.
+    ///
+    /// This is guaranteed to be aligned to the group size.
+    #[inline]
+    pub fn static_empty() -> &'static [u8] {
+        #[repr(C)]
+        struct Dummy {
+            _align: [GroupWord; 0],
+            bytes: [u8; Group::WIDTH],
+        };
+        const DUMMY: Dummy = Dummy {
+            _align: [],
+            bytes: [EMPTY; Group::WIDTH],
+        };
+        &DUMMY.bytes
+    }
+
+    /// Loads a group of bytes starting at the given address.
+    #[inline]
+    pub unsafe fn load(ptr: *const u8) -> Group {
+        Group(ptr::read_unaligned(ptr as *const _))
+    }
+
+    /// Loads a group of bytes starting at the given address, which must be
+    /// aligned to `WIDTH`.
+    #[inline]
+    pub unsafe fn load_aligned(ptr: *const u8) -> Group {
+        Group(ptr::read(ptr as *const _))
```

Comment:
I don't remember if `ptr::read` already performs a `debug_assert!` of the alignment, @RalfJung ? If not (or even if the answer is yes), it might be worth it to add a `debug_assert!` here.

Reply:
Can you try with a `std` `HashSet` instead of `FxHashSet`? If it is a `rustc_resolve` problem, you'll *hopefully* get random test failures.

Reply:
No, it does not. That would be https://github.com/rust-lang/rust/issues/53871.